From f30780f523247d29d55eb3f2fa15a6310864987e Mon Sep 17 00:00:00 2001
From: Bing Song <bing.song@nxp.com>
Date: Fri, 4 Dec 2020 16:31:30 +0800
Subject: [PATCH 30/91] dmabufheaps: memory allocator based on dma-buf heaps.

dma-buf heaps can allocate CMA memory. it can be used for HW.

Upstream-Status: Pending

https://gitlab.freedesktop.org/gstreamer/gst-plugins-base/-/merge_requests/958

Signed-off-by: Bing Song <bing.song@nxp.com>
(cherry picked from commit 48316b86c7cca8006d64b4616a2c17e55927bfe1)
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 gst-libs/gst/allocators/gstdmabufheaps.c | 248 +++++++++++++++++++++++
 gst-libs/gst/allocators/gstdmabufheaps.h |  68 +++++++
 gst-libs/gst/allocators/meson.build      |  12 ++
 3 files changed, 328 insertions(+)
 create mode 100644 gst-libs/gst/allocators/gstdmabufheaps.c
 create mode 100644 gst-libs/gst/allocators/gstdmabufheaps.h

diff --git a/gst-libs/gst/allocators/gstdmabufheaps.c b/gst-libs/gst/allocators/gstdmabufheaps.c
new file mode 100644
index 0000000..69cd45e
--- /dev/null
+++ b/gst-libs/gst/allocators/gstdmabufheaps.c
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2020 NXP
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <linux/dma-buf.h>
+#include <linux/dma-heap.h>
+#include <gst/allocators/gstdmabuf.h>
+#include "gstphysmemory.h"
+#include "gstdmabufheaps.h"
+
+GST_DEBUG_CATEGORY_STATIC (dmabufheaps_allocator_debug);
+#define GST_CAT_DEFAULT dmabufheaps_allocator_debug
+
+#define gst_dmabufheaps_allocator_parent_class parent_class
+
+#define DEFAULT_FD_FLAGS  (O_RDWR | O_CLOEXEC)
+#define DEFAULT_HEAP_FLAGS     DMA_HEAP_VALID_HEAP_FLAGS
+
+enum
+{
+  PROP_0,
+  PROP_FD_FLAGS,
+  PROP_HEAP_FLAGS,
+  PROP_LAST
+};
+
+static guintptr
+gst_dmabufheaps_allocator_get_phys_addr (GstPhysMemoryAllocator * allocator,
+    GstMemory * mem)
+{
+  struct dma_buf_phys dma_phys;
+  gint ret, fd;
+
+  if (!gst_is_dmabuf_memory (mem)) {
+    GST_ERROR ("isn't dmabuf memory");
+    return 0;
+  }
+
+  fd = gst_dmabuf_memory_get_fd (mem);
+  if (fd < 0) {
+    GST_ERROR ("dmabuf memory get fd failed");
+    return 0;
+  }
+
+  GST_DEBUG ("dmabufheaps DMA FD: %d", fd);
+
+  ret = ioctl (fd, DMA_BUF_IOCTL_PHYS, &dma_phys);
+  if (ret < 0)
+    return 0;
+
+  return dma_phys.phys;
+}
+
+static void
+gst_dmabufheaps_allocator_iface_init (gpointer g_iface)
+{
+  GstPhysMemoryAllocatorInterface *iface = g_iface;
+  iface->get_phys_addr = gst_dmabufheaps_allocator_get_phys_addr;
+}
+
+G_DEFINE_TYPE_WITH_CODE (GstDMABUFHEAPSAllocator, gst_dmabufheaps_allocator,
+    GST_TYPE_DMABUF_ALLOCATOR,
+    G_IMPLEMENT_INTERFACE (GST_TYPE_PHYS_MEMORY_ALLOCATOR,
+        gst_dmabufheaps_allocator_iface_init));
+
+static void
+gst_dmabufheaps_allocator_mem_init (void)
+{
+  GstAllocator *allocator =
+      g_object_new (gst_dmabufheaps_allocator_get_type (), NULL);
+  GstDMABUFHEAPSAllocator *self = GST_DMABUFHEAPS_ALLOCATOR (allocator);
+  gint fd;
+
+  fd = open ("/dev/dma_heap/linux,cma", O_RDWR);
+  if (fd < 0) {
+    GST_WARNING ("Could not open dmabufheaps driver");
+    g_object_unref (self);
+    return;
+  }
+
+  self->fd = fd;
+
+  gst_allocator_register (GST_ALLOCATOR_DMABUFHEAPS, allocator);
+}
+
+GstAllocator *
+gst_dmabufheaps_allocator_obtain (void)
+{
+  static GOnce dmabufheaps_allocator_once = G_ONCE_INIT;
+  GstAllocator *allocator;
+
+  g_once (&dmabufheaps_allocator_once,
+      (GThreadFunc) gst_dmabufheaps_allocator_mem_init, NULL);
+
+  allocator = gst_allocator_find (GST_ALLOCATOR_DMABUFHEAPS);
+  if (allocator == NULL)
+    GST_WARNING ("No allocator named %s found", GST_ALLOCATOR_DMABUFHEAPS);
+
+  return allocator;
+}
+
+static GstMemory *
+gst_dmabufheaps_allocator_alloc (GstAllocator * allocator, gsize size,
+    GstAllocationParams * params)
+{
+  GstDMABUFHEAPSAllocator *self = GST_DMABUFHEAPS_ALLOCATOR (allocator);
+  struct dma_heap_allocation_data data = { 0 };
+  GstMemory *mem;
+  gsize dmabufheaps_size;
+  gint dma_fd = -1;
+  gint ret;
+
+  if (self->fd < 0) {
+    GST_WARNING ("don't open dmabufheaps driver");
+    return NULL;
+  }
+
+  dmabufheaps_size = size + params->prefix + params->padding;
+  data.len = dmabufheaps_size,
+      data.fd_flags = self->fd_flags,
+      data.heap_flags = self->heap_flags,
+      ret = ioctl (self->fd, DMA_HEAP_IOCTL_ALLOC, &data);
+  if (ret < 0) {
+    GST_ERROR ("dmabufheaps allocate failed.");
+    return NULL;
+  }
+  dma_fd = data.fd;
+
+  mem =
+      gst_dmabuf_allocator_alloc_with_flags (allocator, dma_fd, size,
+      GST_FD_MEMORY_FLAG_KEEP_MAPPED);
+
+  GST_DEBUG ("dmabufheaps allocated size: %" G_GSIZE_FORMAT "DMA FD: %d",
+      dmabufheaps_size, dma_fd);
+
+  return mem;
+}
+
+static void
+gst_dmabufheaps_allocator_dispose (GObject * object)
+{
+  GstDMABUFHEAPSAllocator *self = GST_DMABUFHEAPS_ALLOCATOR (object);
+
+  if (self->fd > 0) {
+    close (self->fd);
+    self->fd = -1;
+  }
+
+  G_OBJECT_CLASS (parent_class)->dispose (object);
+}
+
+static void
+gst_dmabufheaps_allocator_set_property (GObject * object, guint prop_id,
+    const GValue * value, GParamSpec * pspec)
+{
+  GstDMABUFHEAPSAllocator *self = GST_DMABUFHEAPS_ALLOCATOR (object);
+
+  switch (prop_id) {
+    case PROP_FD_FLAGS:
+      self->fd_flags = g_value_get_uint (value);
+      break;
+    case PROP_HEAP_FLAGS:
+      self->heap_flags = g_value_get_uint (value);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+  }
+}
+
+static void
+gst_dmabufheaps_allocator_get_property (GObject * object, guint prop_id,
+    GValue * value, GParamSpec * pspec)
+{
+  GstDMABUFHEAPSAllocator *self = GST_DMABUFHEAPS_ALLOCATOR (object);
+
+  switch (prop_id) {
+    case PROP_FD_FLAGS:
+      g_value_set_uint (value, self->fd_flags);
+      break;
+    case PROP_HEAP_FLAGS:
+      g_value_set_uint (value, self->heap_flags);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+  }
+}
+
+static void
+gst_dmabufheaps_allocator_class_init (GstDMABUFHEAPSAllocatorClass * klass)
+{
+  GObjectClass *gobject_class = G_OBJECT_CLASS (klass);
+  GstAllocatorClass *allocator_class = GST_ALLOCATOR_CLASS (klass);
+
+  gobject_class->dispose =
+      GST_DEBUG_FUNCPTR (gst_dmabufheaps_allocator_dispose);
+  gobject_class->set_property = gst_dmabufheaps_allocator_set_property;
+  gobject_class->get_property = gst_dmabufheaps_allocator_get_property;
+
+  g_object_class_install_property (gobject_class, PROP_FD_FLAGS,
+      g_param_spec_uint ("fd-flags", "FD Flags",
+          "DMABUFHEAPS fd flags", 0, G_MAXUINT32, DEFAULT_FD_FLAGS,
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_HEAP_FLAGS,
+      g_param_spec_uint ("heap-flags", "Heap Flags",
+          "DMABUFHEAPS heap flags", 0, G_MAXUINT32, DEFAULT_HEAP_FLAGS,
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+
+  allocator_class->alloc = GST_DEBUG_FUNCPTR (gst_dmabufheaps_allocator_alloc);
+
+  GST_DEBUG_CATEGORY_INIT (dmabufheaps_allocator_debug, "dmabufheapsmemory", 0,
+      "DMA FD memory allocator based on dma-buf heaps");
+}
+
+static void
+gst_dmabufheaps_allocator_init (GstDMABUFHEAPSAllocator * self)
+{
+  GstAllocator *allocator = GST_ALLOCATOR (self);
+
+  allocator->mem_type = GST_ALLOCATOR_DMABUFHEAPS;
+
+  self->fd_flags = DEFAULT_FD_FLAGS;
+  self->heap_flags = DEFAULT_HEAP_FLAGS;
+}
diff --git a/gst-libs/gst/allocators/gstdmabufheaps.h b/gst-libs/gst/allocators/gstdmabufheaps.h
new file mode 100644
index 0000000..e5ddba6
--- /dev/null
+++ b/gst-libs/gst/allocators/gstdmabufheaps.h
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2020 NXP
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#ifndef __GST_DMABUFHEAPS_H__
+#define __GST_DMABUFHEAPS_H__
+
+#include <gst/gst.h>
+#include <gst/allocators/gstdmabuf.h>
+#include <gst/allocators/allocators-prelude.h>
+
+G_BEGIN_DECLS
+
+typedef struct _GstDMABUFHEAPSAllocator GstDMABUFHEAPSAllocator;
+typedef struct _GstDMABUFHEAPSAllocatorClass GstDMABUFHEAPSAllocatorClass;
+typedef struct _GstDMABUFHEAPSMemory GstDMABUFHEAPSMemory;
+
+#define GST_ALLOCATOR_DMABUFHEAPS "dmabufheapsmem"
+
+#define GST_TYPE_DMABUFHEAPS_ALLOCATOR gst_dmabufheaps_allocator_get_type ()
+#define GST_IS_DMABUFHEAPS_ALLOCATOR(obj) (G_TYPE_CHECK_INSTANCE_TYPE ((obj), \
+    GST_TYPE_DMABUFHEAPS_ALLOCATOR))
+#define GST_DMABUFHEAPS_ALLOCATOR(obj) \
+  (G_TYPE_CHECK_INSTANCE_CAST((obj), GST_TYPE_DMABUFHEAPS_ALLOCATOR, GstDMABUFHEAPSAllocator))
+#define GST_DMABUFHEAPS_ALLOCATOR_CLASS(klass) \
+  (G_TYPE_CHECK_CLASS_CAST((klass), GST_TYPE_DMABUFHEAPS_ALLOCATOR, GstDMABUFHEAPSAllocatorClass))
+#define GST_DMABUFHEAPS_ALLOCATOR_CAST(obj) ((GstDMABUFHEAPSAllocator *)(obj))
+
+#define GST_DMABUFHEAPS_MEMORY_QUARK gst_dmabufheaps_memory_quark ()
+
+struct _GstDMABUFHEAPSAllocator
+{
+  GstDmaBufAllocator parent;
+
+  gint fd;
+  guint fd_flags;
+  guint heap_flags;
+};
+
+struct _GstDMABUFHEAPSAllocatorClass
+{
+  GstDmaBufAllocatorClass parent;
+};
+
+GST_ALLOCATORS_API
+GType gst_dmabufheaps_allocator_get_type (void);
+
+GST_ALLOCATORS_API
+GstAllocator* gst_dmabufheaps_allocator_obtain (void);
+
+G_END_DECLS
+
+#endif /* __GST_DMABUFHEAPS_H__ */
diff --git a/gst-libs/gst/allocators/meson.build b/gst-libs/gst/allocators/meson.build
index 31f990d..00b5cf2 100644
--- a/gst-libs/gst/allocators/meson.build
+++ b/gst-libs/gst/allocators/meson.build
@@ -17,6 +17,12 @@ if cc.has_header('linux/ion.h')
   ]
 endif
 
+if cc.has_header('linux/dma-heap.h')
+  gst_allocators_headers += [
+    'gstdmabufheaps.h',
+  ]
+endif
+
 install_headers(gst_allocators_headers, subdir : 'gstreamer-1.0/gst/allocators/')
 
 gst_allocators_sources = files([
@@ -53,6 +59,12 @@ if cc.has_header('linux/ion.h')
   ]
 endif
 
+if cc.has_header('linux/dma-heap.h')
+  gst_allocators_sources += [
+    'gstdmabufheaps.c',
+  ]
+endif
+
 gstallocators = library('gstallocators-@0@'.format(api_version),
 
   gst_allocators_sources,
-- 
2.43.0

